src/mono/mono/component/mini-wasi-debugger.c

```diff
@@ -0,0 +1,180 @@
+#include <glib.h>
+#include <mono/mini/debugger-agent-external.h>
+#include <mono/component/debugger-engine.h>
+#include <mono/metadata/components.h>
+
+#ifdef HOST_WASI
+static int conn_fd;
+static int log_level = 1;
+static int retry_receive_message = 100;
+static int connection_wait_us = 1000;
+
+__attribute__((import_module("wasi_snapshot_preview1")))
+__attribute__((import_name("sock_accept")))
+int sock_accept(int fd, int fdflags, int* result_ptr);
+
+static long long
+timeInMilliseconds ()
+{
+	struct timeval tv;
+	gettimeofday(&tv,NULL);
+	return (((long long)tv.tv_sec)*1000)+(tv.tv_usec/1000);
+}
+
+static int
+wasi_transport_recv (void *buf, int len)
+{
+	int res;
+	int total = 0;
+	int fd = conn_fd;
+	int num_recv_calls  = 0;
+
+	do {
+		res = read (fd, (char *) buf + total, len - total);
+		if (res > 0)
+			total += res;
+		num_recv_calls++;
+		if ((res > 0 && total < len) || (res == -1 && num_recv_calls  < retry_receive_message)) {
+			// Wasmtime on Windows doesn't seem to be able to sleep for short periods like 1ms so we'll have to spinlock
+			long long start = timeInMilliseconds ();
+			while (timeInMilliseconds () < start + (connection_wait_us/1000));
+		} else {
+			break;
+		}
+	} while (true);
+	return total;
+}
+
+static gboolean
+wasi_transport_send (void *data, int len)
+{
+	int res;
+	do {
+		res = write (conn_fd, (const char*)data, len);
+	} while (res == -1);
+
+	return res == len;
+}
+
+/*
+ * wasi_transport_connect:
+ *
+ *   listen on the socket that is already opened by wasm runtime
+ */
+static void
+wasi_transport_connect (const char *socket_fd)
```

Comment:
Just curious. Is there way to communicate a "connection error" somewhere? How does it fail for the user, when a connection fails here, or we return without establishing a connection?

Reply:
The way that it works on debugger-agent is if we cannot establish a connection a exit(1) is called.

https://github.com/dotnet/runtime/blob/6ee429c704e52c48cee7902f76df7b358f8c19ec/src/mono/mono/component/debugger-agent.c#L1217

We can to the same after some retries, @SteveSandersonMS , do you want this?

Reply:
I'd say if we know objectively that the connection has failed then `exit(1)` makes sense. However if we're possibly still waiting for a debugger to connect (even after a long time) it's probably better to keep waiting.